### PR TITLE
Add support for API Key management

### LINKED ISSUE #70: [Feature] Support for API Key Management
**Is your feature request related to a problem? Please describe.**
Managing Elasticsearch API keys is somewhat cumbersome, especially on large scale deployments where you want fine grain permission controls.

**Describe the resource you would like to have implemented.**
Ability to manage/create API keys via Terraform.

**Describe the solution you'd like**
I'd like to be able to specify an API key resource:
- For creation and be able to get back the information needed for using the API key.
- That exists and delete it via terraform.

I'd like to be able to get a list of existing API keys via a data source.

**Describe alternatives you've considered**
You can use the Elasticsearch API via something like curl:
- Create API Key https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html
  - Grant API Key https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-grant-api-key.html
- Delete API Key https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-invalidate-api-key.html
- API Key Data Source https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-api-key.html
- API Key List Data Source https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html

**Additional context**
Security Resources are already partially supported via [user](https://registry.terraform.io/providers/elastic/elasticstack/latest/docs/resources/elasticsearch_security_user) and [role](https://registry.terraform.io/providers/elastic/elasticstack/latest/docs/resources/elasticsearch_security_role)

### PR DESCRIPTION
**Pull request by webfella**
Implements the ability to manage/create API keys via Terraform, based on [some of the initial work done here](https://github.com/jitbasemartin/terraform-provider-elasticstack/tree/feature/issue-70).

Closes #70.

### CODE REVIEW

File: internal/clients/security.go [hunk 1]
```diff
@@ -195,3 +195,93 @@ func (a *ApiClient) DeleteElasticsearchRoleMapping(ctx context.Context, roleMapp
...
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+
+	res, err := a.es.Security.CreateAPIKey(bytes.NewReader(apikeyBytes))
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if diags := utils.CheckError(res, "Unable to create or update an apikey"); diags.HasError() {
```
Comment (tobio) [inline internal/clients/security.go]:
```suggestion
	if diags := utils.CheckError(res, "Unable to create apikey"); diags.HasError() {
```

File: internal/clients/security.go [hunk 2]
```diff
@@ -195,3 +195,93 @@ func (a *ApiClient) DeleteElasticsearchRoleMapping(ctx context.Context, roleMapp
...
+	req := a.es.Security.GetAPIKey.WithID(id)
+	res, err := a.es.Security.GetAPIKey(req)
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if res.StatusCode == http.StatusNotFound {
+		diags := append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to find a apikey in the cluster.",
```
Comment (tobio) [inline internal/clients/security.go]:
```suggestion
			Summary:  "Unable to find an apikey in the cluster.",
```

File: internal/clients/security.go [hunk 3]
```diff
@@ -195,3 +195,93 @@ func (a *ApiClient) DeleteElasticsearchRoleMapping(ctx context.Context, roleMapp
...
+			Detail:   fmt.Sprintf("Unable to get apikey: '%s' from the cluster.", id),
+		})
+		return nil, diags
+	}
+	if diags := utils.CheckError(res, "Unable to get a apikey."); diags.HasError() {
```
Comment (tobio) [inline internal/clients/security.go]:
```suggestion
	if diags := utils.CheckError(res, "Unable to get an apikey."); diags.HasError() {
```

File: internal/clients/security.go [hunk 4]
```diff
@@ -195,3 +195,93 @@ func (a *ApiClient) DeleteElasticsearchRoleMapping(ctx context.Context, roleMapp
...
+		ApiKeys []models.ApiKeyResponse `json:"api_keys"`
+	}
+	if err := json.NewDecoder(res.Body).Decode(&apiKeys); err != nil {
+		return nil, diag.FromErr(err)
+	}
+
+	if len(apiKeys.ApiKeys) != 1 {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Unable to find a apikey in the cluster",
```
Comment (tobio) [inline internal/clients/security.go]:
```suggestion
			Summary:  "Unable to find an apikey in the cluster",
```

File: internal/clients/security.go [hunk 5]
```diff
@@ -195,3 +195,93 @@ func (a *ApiClient) DeleteElasticsearchRoleMapping(ctx context.Context, roleMapp
...
+	apikeyBytes, err := json.Marshal(apiKeys)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	res, err := a.es.Security.InvalidateAPIKey(bytes.NewReader(apikeyBytes))
+	if err != nil && res.IsError() {
+		return diag.FromErr(err)
+	}
+	defer res.Body.Close()
+	if diags := utils.CheckError(res, "Unable to delete a apikey"); diags.HasError() {
```
Comment (tobio) [inline internal/clients/security.go]:
```suggestion
	if diags := utils.CheckError(res, "Unable to delete an apikey"); diags.HasError() {
```

File: internal/elasticsearch/security/api_key.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+	if v, ok := d.GetOk("role_descriptors"); ok {
+		role_descriptors := map[string]models.Role{}
+		api_role_descriptors := map[string]models.ApiKeyRole{}
+
+		if err := json.NewDecoder(strings.NewReader(v.(string))).Decode(&role_descriptors); err != nil {
+			return diag.FromErr(err)
+		}
+
+		for key, value := range role_descriptors {
```
Comment (tobio) [inline internal/elasticsearch/security/api_key.go]: Can we add a comment here about why we need this nonsense.

File: internal/elasticsearch/security/api_key.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		if err := d.Set("api_key", putResponse.Key); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	if putResponse.EncodedKey != "" {
+		if err := d.Set("encoded", putResponse.EncodedKey); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
```
Comment (tobio) [inline internal/elasticsearch/security/api_key.go]:
```suggestion
```

nit

**Review by tobio — COMMENTED** [review]
Couple of minor comments.

File: internal/elasticsearch/security/api_key.go (hunk 6 above)
Comment (webfella) [reply internal/elasticsearch/security/api_key.go]: Verified it works when removing the index field ⚡

File: internal/elasticsearch/security/api_key_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+							return fmt.Errorf("%v doesn't match %v", testRoleDescriptor, expectedRoleDescriptor)
+						}
+
+						return nil
+					}),
+					resource.TestCheckResourceAttrSet("elasticstack_elasticsearch_security_api_key.test", "expiration"),
+					resource.TestCheckResourceAttrSet("elasticstack_elasticsearch_security_api_key.test", "api_key"),
+					resource.TestCheckResourceAttrSet("elasticstack_elasticsearch_security_api_key.test", "encoded"),
+				),
+				SkipFunc: checkIfVersionIsUnsupported,
```
Comment (tobio) [inline internal/elasticsearch/security/api_key_test.go]: nit: Can we put this at the top of the step, IMO it's easy to miss once it's mixed in with all the checks.

**Review by tobio — COMMENTED** [review]
Being a pedant, sorry.

File: internal/elasticsearch/security/api_key_test.go (hunk 8 above)
Comment (webfella) [reply internal/elasticsearch/security/api_key_test.go]: All good, I had it up there and decided i'd alphabetise it for some reason! The more pedantic the better for me at the moment mate 👍

**Review by tobio — APPROVED** [review]

**Comment by jitbasemartin** [comment]
Thanks @webfella !

**Pull request merged**